File: src/main/java/com/library/system/LibrarySystem.java
package com.library.system;

import com.library.models.Livre;
import com.library.models.TransactionEmprunt;
import com.library.users.Utilisateur;
import com.library.observers.NotificationService;
import java.time.LocalDate;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.stream.Collectors;

/**
 * Singleton - LibrarySystem
 * Système principal de gestion de la bibliothèque
 * Assure qu'une seule instance existe dans toute l'application
 */
public class LibrarySystem {
    private static LibrarySystem instance;
    
    private Map<String, Utilisateur> utilisateurs;
    private Map<String, Livre> livres;
    private List<TransactionEmprunt> transactions;
    private NotificationService notificationService;
    private int compteurTransactions;

    /**
     * Constructeur privé pour empêcher l'instanciation directe
     */
    private LibrarySystem() {
        this.utilisateurs = new HashMap<>();
        this.livres = new HashMap<>();
        this.transactions = new ArrayList<>();
        this.notificationService = new NotificationService();
        this.compteurTransactions = 0;
    }

    /**
     * Retourne l'instance unique de LibrarySystem
     * @return L'instance du système de bibliothèque
     */
    public static synchronized LibrarySystem getInstance() {
        if (instance == null) {
            instance = new LibrarySystem();
        }
        return instance;
    }

    // ========== GESTION DES UTILISATEURS ==========

    public void addUtilisateur(Utilisateur utilisateur) {
        utilisateurs.put(utilisateur.getId(), utilisateur);
        notificationService.subscribe(utilisateur);
        System.out.println("✓ Utilisateur ajouté: " + utilisateur.getNom());
    }

    public Utilisateur getUtilisateur(String id) {
        return utilisateurs.get(id);
    }

    public List<Utilisateur> getAllUtilisateurs() {
        return new ArrayList<>(utilisateurs.values());
    }

    public void removeUtilisateur(String id) {
        Utilisateur user = utilisateurs.remove(id);
        if (user != null) {
            notificationService.unsubscribe(user);
            System.out.println("✓ Utilisateur supprimé: " + user.getNom());
        }
    }

    // ========== GESTION DES LIVRES ==========

    public void addLivre(Livre livre) {
        livres.put(livre.getIsbn(), livre);
        System.out.println("✓ Livre ajouté: " + livre.getTitre());
    }

    public Livre getLivre(String isbn) {
        return livres.get(isbn);
    }

    public List<Livre> getAllLivres() {
        return new ArrayList<>(livres.values());
    }

    public List<Livre> getLivresDisponibles() {
        return livres.values().stream()
                .filter(Livre::isDisponible)
                .collect(Collectors.toList());
    }

    public void removeLivre(String isbn) {
        Livre livre = livres.remove(isbn);
        if (livre != null) {
            System.out.println("✓ Livre supprimé: " + livre.getTitre());
        }
    }

    // ========== GESTION DES EMPRUNTS ==========

    public boolean emprunterLivre(String idUtilisateur, String isbnLivre) {
        Utilisateur utilisateur = getUtilisateur(idUtilisateur);
        Livre livre = getLivre(isbnLivre);

        if (utilisateur == null) {
            System.out.println("✗ Erreur: Utilisateur introuvable");
            return false;
        }

        if (livre == null) {
            System.out.println("✗ Erreur: Livre introuvable");
            return false;
        }

        if (!livre.isDisponible()) {
            System.out.println("✗ Erreur: Le livre '" + livre.getTitre() + "' n'est pas disponible");
            return false;
        }

        if (!utilisateur.peutEmprunter()) {
            System.out.println("✗ Erreur: " + utilisateur.getNom() + " a atteint sa limite d'emprunts");
            return false;
        }

        // Créer la transaction
        compteurTransactions++;
        String idTransaction = "TRANS-" + compteurTransactions;
        TransactionEmprunt transaction = new TransactionEmprunt(idTransaction, idUtilisateur, isbnLivre);
        transactions.add(transaction);

        // Mettre à jour l'état du livre
        livre.setDisponible(false);
        livre.setEmpruntePar(idUtilisateur);

        // Mettre à jour l'utilisateur
        utilisateur.addLivreEmpruntu(livre.getTitre());

        // Notification
        String message = utilisateur.getNom() + " a emprunté '" + livre.getTitre() + 
                        "'. Retour prévu le: " + transaction.getDateRetourPrevue();
        notificationService.notifyObservers(message);

        System.out.println("✓ Emprunt réussi: " + utilisateur.getNom() + " a emprunté '" + livre.getTitre() + "'");
        return true;
    }

    public boolean retournerLivre(String idUtilisateur, String isbnLivre) {
        Utilisateur utilisateur = getUtilisateur(idUtilisateur);
        Livre livre = getLivre(isbnLivre);

        if (utilisateur == null || livre == null) {
            System.out.println("✗ Erreur: Utilisateur ou livre introuvable");
            return false;
        }

        // Trouver la transaction active
        TransactionEmprunt transaction = transactions.stream()
                .filter(t -> t.getIdUtilisateur().equals(idUtilisateur) &&
                        t.getIsbnLivre().equals(isbnLivre) &&
                        !t.isRetourne())
                .findFirst()
                .orElse(null);

        if (transaction == null) {
            System.out.println("✗ Erreur: Aucun emprunt actif trouvé pour ce livre");
            return false;
        }

        // Mettre à jour la transaction
        transaction.setDateRetourEffective(LocalDate.now());

        // Mettre à jour le livre
        livre.liberer();

        // Mettre à jour l'utilisateur
        utilisateur.removeLivreEmpruntu(livre.getTitre());

        // Notification
        String message;
        if (transaction.isEnRetard()) {
            long joursRetard = transaction.getJoursDeRetard();
            message = utilisateur.getNom() + " a retourné '" + livre.getTitre() + 
                    "' avec " + joursRetard + " jours de retard!";
        } else {
            message = utilisateur.getNom() + " a retourné '" + livre.getTitre() + "' à temps.";
        }
        notificationService.notifyObservers(message);

        System.out.println("✓ Retour effectué: " + message);
        return true;
    }

    public List<TransactionEmprunt> getTransactionsUtilisateur(String idUtilisateur) {
        return transactions.stream()
                .filter(t -> t.getIdUtilisateur().equals(idUtilisateur))
                .collect(Collectors.toList());
    }

    public List<TransactionEmprunt> getTransactionsEnRetard() {
        return transactions.stream()
                .filter(t -> t.isEnRetard() && !t.isRetourne())
                .collect(Collectors.toList());
    }

    // ========== NOTIFICATIONS ==========

    public void notifierRetards() {
        List<TransactionEmprunt> retards = getTransactionsEnRetard();
        for (TransactionEmprunt transaction : retards) {
            Utilisateur utilisateur = getUtilisateur(transaction.getIdUtilisateur());
            Livre livre = getLivre(transaction.getIsbnLivre());
            if (utilisateur != null && livre != null) {
                String message = "⚠️ ALERTE RETARD: Vous devez retourner '" + livre.getTitre() + 
                                "' depuis " + transaction.getJoursDeRetard() + " jours!";
                notificationService.notifyObservers(message);
            }
        }
    }

    // ========== AFFICHAGE ET RAPPORTS ==========

    public void afficherStatistiques() {
        System.out.println("\n========== STATISTIQUES DE LA BIBLIOTHÈQUE ==========");
        System.out.println("Utilisateurs: " + utilisateurs.size());
        System.out.println("Livres: " + livres.size());
        System.out.println("Livres disponibles: " + getLivresDisponibles().size());
        System.out.println("Transactions totales: " + transactions.size());
        System.out.println("Emprunts en retard: " + getTransactionsEnRetard().size());
    }

    public void afficherLivresEmpruntés(String idUtilisateur) {
        Utilisateur utilisateur = getUtilisateur(idUtilisateur);
        if (utilisateur != null) {
            System.out.println("\nLivres empruntés par " + utilisateur.getNom() + ":");
            if (utilisateur.getLivresEmpruntés().isEmpty()) {
                System.out.println("  Aucun livre emprunté");
            } else {
                utilisateur.getLivresEmpruntés().forEach(livre -> System.out.println("  - " + livre));
            }
        }
    }
}

File: js/src/system/LibrarySystem.js
const Livre = require('../models/Livre');
const TransactionEmprunt = require('../models/TransactionEmprunt');
const NotificationService = require('../observers/NotificationService');

/**
 * Singleton - LibrarySystem
 * Système principal de gestion de la bibliothèque
 * Assure qu'une seule instance existe dans toute l'application
 */
class LibrarySystem {
    constructor() {
        if (LibrarySystem.instance) {
            return LibrarySystem.instance;
        }

        this.utilisateurs = new Map();
        this.livres = new Map();
        this.transactions = [];
        this.notificationService = new NotificationService();
        this.compteurTransactions = 0;

        LibrarySystem.instance = this;
    }

    /**
     * Retourne l'instance unique de LibrarySystem
     * @returns {LibrarySystem} L'instance du système de bibliothèque
     */
    static getInstance() {
        if (!LibrarySystem.instance) {
            new LibrarySystem();
        }
        return LibrarySystem.instance;
    }

    // ========== GESTION DES UTILISATEURS ==========

    /**
     * Ajouter un utilisateur
     * @param {Utilisateur} utilisateur - L'utilisateur à ajouter
     */
    addUtilisateur(utilisateur) {
        this.utilisateurs.set(utilisateur.getId(), utilisateur);
        this.notificationService.subscribe(utilisateur);
        console.log(`✓ Utilisateur ajouté: ${utilisateur.getNom()}`);
    }

    /**
     * Obtenir un utilisateur par ID
     * @param {string} id - L'ID de l'utilisateur
     * @returns {Utilisateur} L'utilisateur ou undefined
     */
    getUtilisateur(id) {
        return this.utilisateurs.get(id);
    }

    /**
     * Obtenir tous les utilisateurs
     * @returns {Array} Liste des utilisateurs
     */
    getAllUtilisateurs() {
        return Array.from(this.utilisateurs.values());
    }

    /**
     * Supprimer un utilisateur
     * @param {string} id - L'ID de l'utilisateur
     */
    removeUtilisateur(id) {
        const user = this.utilisateurs.get(id);
        if (user) {
            this.utilisateurs.delete(id);
            this.notificationService.unsubscribe(user);
            console.log(`✓ Utilisateur supprimé: ${user.getNom()}`);
        }
    }

    // ========== GESTION DES LIVRES ==========

    /**
     * Ajouter un livre
     * @param {Livre} livre - Le livre à ajouter
     */
    addLivre(livre) {
        this.livres.set(livre.getIsbn(), livre);
        console.log(`✓ Livre ajouté: ${livre.getTitre()}`);
    }

    /**
     * Obtenir un livre par ISBN
     * @param {string} isbn - L'ISBN du livre
     * @returns {Livre} Le livre ou undefined
     */
    getLivre(isbn) {
        return this.livres.get(isbn);
    }

    /**
     * Obtenir tous les livres
     * @returns {Array} Liste des livres
     */
    getAllLivres() {
        return Array.from(this.livres.values());
    }

    /**
     * Obtenir les livres disponibles
     * @returns {Array} Liste des livres disponibles
     */
    getLivresDisponibles() {
        return this.getAllLivres().filter(livre => livre.isDisponible());
    }

    /**
     * Supprimer un livre
     * @param {string} isbn - L'ISBN du livre
     */
    removeLivre(isbn) {
        const livre = this.livres.get(isbn);
        if (livre) {
            this.livres.delete(isbn);
            console.log(`✓ Livre supprimé: ${livre.getTitre()}`);
        }
    }

    // ========== GESTION DES EMPRUNTS ==========

    /**
     * Emprunter un livre
     * @param {string} idUtilisateur - L'ID de l'utilisateur
     * @param {string} isbnLivre - L'ISBN du livre
     * @returns {boolean} True si l'emprunt a réussi
     */
    emprunterLivre(idUtilisateur, isbnLivre) {
        const utilisateur = this.getUtilisateur(idUtilisateur);
        const livre = this.getLivre(isbnLivre);

        if (!utilisateur) {
            console.log('✗ Erreur: Utilisateur introuvable');
            return false;
        }

        if (!livre) {
            console.log('✗ Erreur: Livre introuvable');
            return false;
        }

        if (!livre.isDisponible()) {
            console.log(`✗ Erreur: Le livre '${livre.getTitre()}' n'est pas disponible`);
            return false;
        }

        if (!utilisateur.peutEmprunter()) {
            console.log(`✗ Erreur: ${utilisateur.getNom()} a atteint sa limite d'emprunts`);
            return false;
        }

        // Créer la transaction
        this.compteurTransactions++;
        const idTransaction = `TRANS-${this.compteurTransactions}`;
        const transaction = new TransactionEmprunt(idTransaction, idUtilisateur, isbnLivre);
        this.transactions.push(transaction);

        // Mettre à jour l'état du livre
        livre.setDisponible(false);
        livre.setEmpruntePar(idUtilisateur);

        // Mettre à jour l'utilisateur
        utilisateur.addLivreEmpruntu(livre.getTitre());

        // Notification
        const message = `${utilisateur.getNom()} a emprunté '${livre.getTitre()}'. Retour prévu le: ${TransactionEmprunt.formatDate(transaction.getDateRetourPrevue())}`;
        this.notificationService.notifyObservers(message);

        console.log(`✓ Emprunt réussi: ${utilisateur.getNom()} a emprunté '${livre.getTitre()}'`);
        return true;
    }

    /**
     * Retourner un livre
     * @param {string} idUtilisateur - L'ID de l'utilisateur
     * @param {string} isbnLivre - L'ISBN du livre
     * @returns {boolean} True si le retour a réussi
     */
    retournerLivre(idUtilisateur, isbnLivre) {
        const utilisateur = this.getUtilisateur(idUtilisateur);
        const livre = this.getLivre(isbnLivre);

        if (!utilisateur || !livre) {
            console.log('✗ Erreur: Utilisateur ou livre introuvable');
            return false;
        }

        // Trouver la transaction active
        const transaction = this.transactions.find(t =>
            t.getIdUtilisateur() === idUtilisateur &&
            t.getIsbnLivre() === isbnLivre &&
            !t.isRetourne()
        );

        if (!transaction) {
            console.log('✗ Erreur: Aucun emprunt actif trouvé pour ce livre');
            return false;
        }

        // Mettre à jour la transaction
        transaction.setDateRetourEffective(new Date());

        // Mettre à jour le livre
        livre.liberer();

        // Mettre à jour l'utilisateur
        utilisateur.removeLivreEmpruntu(livre.getTitre());

        // Notification
        let message;
        if (transaction.isEnRetard()) {
            const joursRetard = transaction.getJoursDeRetard();
            message = `${utilisateur.getNom()} a retourné '${livre.getTitre()}' avec ${joursRetard} jours de retard!`;
        } else {
            message = `${utilisateur.getNom()} a retourné '${livre.getTitre()}' à temps.`;
        }
        this.notificationService.notifyObservers(message);

        console.log(`✓ Retour effectué: ${message}`);
        return true;
    }

    /**
     * Obtenir les transactions d'un utilisateur
     * @param {string} idUtilisateur - L'ID de l'utilisateur
     * @returns {Array} Liste des transactions
     */
    getTransactionsUtilisateur(idUtilisateur) {
        return this.transactions.filter(t => t.getIdUtilisateur() === idUtilisateur);
    }

    /**
     * Obtenir les transactions en retard
     * @returns {Array} Liste des transactions en retard
     */
    getTransactionsEnRetard() {
        return this.transactions.filter(t => t.isEnRetard() && !t.isRetourne());
    }

    // ========== NOTIFICATIONS ==========

    /**
     * Notifier les utilisateurs en retard
     */
    notifierRetards() {
        const retards = this.getTransactionsEnRetard();
        retards.forEach(transaction => {
            const utilisateur = this.getUtilisateur(transaction.getIdUtilisateur());
            const livre = this.getLivre(transaction.getIsbnLivre());
            if (utilisateur && livre) {
                const message = `⚠️ ALERTE RETARD: Vous devez retourner '${livre.getTitre()}' depuis ${transaction.getJoursDeRetard()} jours!`;
                this.notificationService.notifyObservers(message);
            }
        });
    }

    // ========== AFFICHAGE ET RAPPORTS ==========

    /**
     * Afficher les statistiques de la bibliothèque
     */
    afficherStatistiques() {
        console.log('\n========== STATISTIQUES DE LA BIBLIOTHÈQUE ==========');
        console.log(`Utilisateurs: ${this.utilisateurs.size}`);
        console.log(`Livres: ${this.livres.size}`);
        console.log(`Livres disponibles: ${this.getLivresDisponibles().length}`);
        console.log(`Transactions totales: ${this.transactions.length}`);
        console.log(`Emprunts en retard: ${this.getTransactionsEnRetard().length}`);
    }

    /**
     * Afficher les livres empruntés par un utilisateur
     * @param {string} idUtilisateur - L'ID de l'utilisateur
     */
    afficherLivresEmpruntés(idUtilisateur) {
        const utilisateur = this.getUtilisateur(idUtilisateur);
        if (utilisateur) {
            console.log(`\nLivres empruntés par ${utilisateur.getNom()}:`);
            const livres = utilisateur.getLivresEmpruntés();
            if (livres.length === 0) {
                console.log('  Aucun livre emprunté');
            } else {
                livres.forEach(livre => console.log(`  - ${livre}`));
            }
        }
    }
}

module.exports = LibrarySystem;
